platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,762 @@
+
+        Self {
+            smhc,
+            isr: &SMHC0_ISR,
+            int: (int, isr),
+        }
+    }
+
+    /// # Safety
+    /// The descriptor chain needs to live at least as long as the DMA transfer.
+    /// Additionally, their content (e.g., the validity of the buffers they point to)
+    /// also needs to be verified by the user.
+    unsafe fn prepare_dma(&self, descriptor: &idmac::Descriptor, byte_cnt: u32) {
+        self.smhc.smhc_ctrl.modify(|_, w| {
+            w.dma_enb().set_bit();
+            w.dma_rst().set_bit();
+            w
+        });
+        while self.smhc.smhc_ctrl.read().dma_rst().bit_is_set() {}
+
+        // Configure the address of the first DMA descriptor
+        // Right-shift by 2 because it is a *word-address*.
+        self.smhc
+            .smhc_dlba
+            .write(|w| unsafe { w.bits((descriptor as *const _ as u32) >> 2) });
+
+        // Set number of bytes that will be read or written.
+        self.smhc.smhc_bytcnt.write(|w| unsafe { w.bits(byte_cnt) });
+
+        // Soft reset of DMA controller
+        self.smhc.smhc_idmac.write(|w| w.idmac_rst().set_bit());
+
+        // Configure the burst size and TX/RX trigger level
+        self.smhc.smhc_fifoth.write(|w| {
+            w.tx_tl().variant(8);
+            w.rx_tl().variant(7);
+            w.bsize_of_trans().t8();
+            w
+        });
+
+        // Configure the transfer interrupt, receive interrupt, and abnormal interrupt.
+        self.smhc.smhc_idie.write(|w| {
+            w.rx_int_enb().set_bit();
+            w.tx_int_enb().set_bit();
+            w.err_sum_int_enb().set_bit();
+            w
+        });
+
+        // Enable the IDMAC and configure burst transfers
+        self.smhc.smhc_idmac.write(|w| {
+            w.idmac_enb().set_bit();
+            w.fix_bust_ctrl().set_bit();
+            w
+        });
+
+        // Reset the FIFO
+        self.smhc.smhc_ctrl.modify(|_, w| w.fifo_rst().reset());
+        while self.smhc.smhc_ctrl.read().fifo_rst().is_reset() {
+            core::hint::spin_loop();
+        }
+    }
+
+    pub fn handle_smhc0_interrupt() {
+        let _isr = kernel::isr::Isr::enter();
+        let smhc = unsafe { &*SMHC0::ptr() };
+        // safety: it's okay to do this since this function can only be called from inside the ISR.
+        let data = unsafe { &mut (*SMHC0_ISR.data.get()) };
+
+        data.advance_isr(smhc, 0);
+    }
+
+    pub async fn register(
+        self,
+        kernel: &'static Kernel,
+        queued: usize,
+    ) -> Result<(), registry::RegistrationError> {
+        let rx = kernel
+            .registry()
+            .bind_konly::<SdmmcService>(queued)
+            .await?
+            .into_request_stream(queued)
+            .await;
+
+        kernel.spawn(self.run(rx)).await;
+        tracing::info!("SMHC driver task spawned");
+
+        Ok(())
+    }
+
+    #[tracing::instrument(name = "SMHC", level = tracing::Level::INFO, skip(self, rx))]
```

Comment:
tiny nit: can we include the SMHC peripheral number in this span, so we can tell whether it's SMHC0, SMHC1, or SMHC2?

Reply:
That would be good. I guess the `Smhc` struct should then also include its *number* (0, 1, 2) to make this possible?

Reply:
yeah, adding a u8 for it seems fine to me!